odl/trafos/wavelet.py

```diff
@@ -834,14 +957,511 @@ def adjoint(self):
+                                                wbasis=self.wbasis)
+
+
+class AdjBiorthWaveletTransform(Operator):
+    """Discrete adjoint of biorthogonal wavelet transform between L2 space."""
+
+    def __init__(self, ran, nscales, wbasis):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        ran : `DiscreteLpVector` ("image domain")
+
+        nscales : `int`
+            Number of scaling levels in the wavelet transform.
+
+        wbasis : `str`
+            (``'JOSbiorthN'``) J.-O. Stomberg's biorthogonal wavelets
+            where ``N`` refers to the filter length.
+            Possible values are:
+
+            1, 3, 5, 7, 9
+
+        Signal extension mode: reflexion
+        """
+
+        self.nscales = int(nscales)
+        self.wbasis = wbasis
+
+        if not isinstance(ran, DiscreteLp):
+            raise TypeError('domain {!r} is not a `DiscreteLp` instance.'
+                            ''.format(ran))
+
+        if ran.exponent != 2.0:
+            raise ValueError('domain Lp exponent is {} instead of 2.0.'
+                             ''.format(ran.exponent))
+
+        ndim = ran.ndim
+        if ndim == 1:
+            max_level = np.log2(ran.grid.shape[0])
+        elif ndim == 2:
+            max_level = np.log2(max(ran.grid.shape[0], ran.grid.shape[1]))
+        elif ndim == 3:
+            max_level = np.log2(max(ran.grid.shape[0], ran.grid.shape[1],
+                                ran.grid.shape[2]))
+        else:
+            raise NotImplementedError('ndim {} not 1, 2 or 3'
+                                      ''.format(len(ran.ndim)))
+
+        max_level = np.ceil(max_level)
+        if nscales > max_level:
+            raise ValueError('Maximum useful number of scaling levels is {}, '
+                             'got {}.'.format(max_level, self.nscales))
+            #raiseETC ('Setting {} to {}' .format(self.nscales, max_level))
+
+        filterlength = int(wbasis[9])
+        if filterlength not in (1, 3, 5, 7, 9):
+            raise NotImplementedError('Filterlength {} not 1, 3, 5, 7 or 9'
+                                      ''.format(filterlength))
+
+        dom_size = ran.size
+        dom = ran.dspace_type(dom_size, dtype=ran.dtype)
+        super().__init__(dom, ran, linear=True)
+
+    def _call(self, coeff):
+        """Discrete adjoint wavelet transform with biorthogonal wavelets.
+
+        Parameters
+        ----------
+        coeff : `DiscreteLpVector`
+
+        Returns
+        -------
+        arr : `DiscreteLpVector`
+        """
+        filterlength = int(self.wbasis[9])
+        if len(self.range.grid.shape) == 1:
+            nx = self.range.grid.shape[0]
+            #A copy of coeff is unavoidable since bwt writes over
+            #the given input
+            coeff_in = coeff.copy()
+            coeff_in = coeff_in.asarray()
+            x = np.asarray(self.range.element())
```

Comment:
Decide on either `vector.asarray()` or `np.asarray(vector)`, but try to be consistent.